### PR TITLE
0002 ckb whitepaper major update

### PR DESCRIPTION
**Pull request by janx**
This is a major update to CKB whitepaper, one year after its publication. I added the latest results come from our discussions and developments and removed obsolete contents.

[direct link](https://github.com/nervosnetwork/rfcs/blob/deb4fa72de83a6358cb79ae859151c484699f8c8/rfcs/0002-ckb/0002-ckb.md)

### CODE REVIEW

File: rfcs/0002-ckb/0002-ckb.md [hunk 1]
```diff
@@ -7,372 +7,232 @@ Organization: Nervos Foundation
...
+CKB provides a stateful Turing-complete programming model based on CKB VM and cell model.
 
-The design of the CKB cell model and transactions is friendly to light clients. Since all the states are in blocks, block synchronization also accomplishes state synchronization. Light clients only need to synchronize blocks, and do not need to perform state transition computations. If we only stored events in blocks, we would have needed full nodes to also support state synchronization. This extra protocol can be difficult for large deployments, because the incentive to do so is not defined within the blockchain protocol. CKB defines state synchronization in the protocol itself, and this makes light nodes and full nodes more equal peers, leading to a more robust and decentralized system.
+| | Bitcoin | Ethereum | CKB |
+|-|---------|----------|------------|
+|Instruction Set|Script|EVM|RISC-V|
+|Cryptographic Primitive|Opcode|Precompile|Syscall|
```
Comment (xxuejie) [inline rfcs/0002-ckb/0002-ckb.md]: I think we haven't reached the point where we need to use syscall to implement crypto primitives, correct?

Comment (janx) [reply rfcs/0002-ckb/0002-ckb.md]: Correct, 'Syscall' here is wrong. I was thinking of other things when writing this. Fixed.

File: rfcs/0002-ckb/0002-ckb.md [hunk 2]
```diff
@@ -7,372 +7,234 @@ Organization: Nervos Foundation
...
 
-Cell is a more generic storage model compared to the UTXO or the account model. Both the UTXO model and the account model can express relationships between assets and their owners. The UTXO model defines ownership on assets (with the lock script), while the account model defines ownership of assets on owners (with the balance). The UTXO model makes the ledger history more clear, but its lack of explicit accounts makes its already inexpressive scripts harder to use. There is also no way to store account metadata such as authorizations conveniently. The account model is easy to understand, and can support authorizations and identities well, but is not easy to process transactions in parallel. The Cell model, with types and identity, takes the best of both models to provide a more generic data model.
+Correctness includes eventual consistency that guarantees every node see identical copies of state and availability that makes sure the network respond to user’s requests within a reasonable time, and fairness that ensures mining nodes to get fair returns for their efforts to keep the network securely functioning.
```
Comment (doitian) [inline rfcs/0002-ckb/0002-ckb.md]:
```suggestion
Correctness includes eventual consistency, availability, and fairness. Eventual consistency guarantees every node see identical copies of state. Availability makes sure the network respond to user’s requests within a reasonable time. Fairness ensures mining nodes to get fair returns for their efforts to keep the network securely functioning.
```

It is much clear to list all the three factors first.

Comment (doitian) [reply rfcs/0002-ckb/0002-ckb.md]: This paragraph is a bit strange here. It just explains what is correctness, but what's the relationship between it and CKB consensus?

Comment (janx) [reply rfcs/0002-ckb/0002-ckb.md]: > CKB consensus is an improved Nakamoto consensus based on Proof-of-Work, it aims to achieve openness, correctness and high performance in distributed environments with network delay and Byzantine node faults.

File: rfcs/0002-ckb/0002-ckb.md [hunk 3]
```diff
@@ -7,372 +7,234 @@ Organization: Nervos Foundation
...
 
-The creation and storage of states on the CKB incur cost. The creation of states needs to be validated by full nodes, incurring computational cost; the storage of states needs full nodes to provide storage space on an ongoing basis. Current permissionless blockchains only charge one time transaction fees, but they allow data to be stored on all full nodes, occupying storage space forever.
+In CKB, cells are basic storage units of states. Unoccupied cell capacity is transferable, and this gives cells liquidity. Users cannot transfer occupied cell capacity and lose liquidity on their occupied capacity. Therefore, cell owners pay for storage with the liquidity loss of cell capacity, which is an opportunity cost. The larger capacity and the longer time they occupy, the higher liquidity cost they pay. The advantage of pay by liquidity loss, compared to upfront payments, is that it avoids the problem that upfront payments could be used up, and the system would have to recycle the cells. Cells could have different owners and users, and owners can pay the liquidity cost on behalf of their users.
```
Comment (doitian) [inline rfcs/0002-ckb/0002-ckb.md]: > Users cannot transfer occupied cell capacity and lose liquidity on their occupied capacity.

This sentence makes me confused. The two clauses indeed are a cause-result relationship, so **thus** is a better choice than **and**.

Strictly speaking, users do be able to transfer occupied cell capacity. What users cannot is to transfer without losing the state.

File: rfcs/0002-ckb/0002-ckb.md [hunk 4]
```diff
@@ -7,372 +7,234 @@ Organization: Nervos Foundation
...
+- `capacity` - Size limit of this cell. A cell's size is the total size of all fields in it.
+- `data` - State data stored in this cell. It could be empty. Total bytes used by a cell, including data, must always be less than or equal to its capacity.
+- `type`: State verification script.
+- `lock`: Script that represents the ownership of the cell. Owners of cells can transfer cells to others.
 
-  The generation algorithms can be implemented in any programming languages, using the CKB light client API or the CKB SDK.
+A cell is an immutable object, as no one can modify it after creation. Every cell can only be used once - it cannot be used as inputs for two different transactions. Cell ‘updates’ mark previous cells as history and create new cells with identity capacity to replace them. By construct and send transactions, users provide new cells with new states in it and invalidate previous cells storing old states atomically. The set of all current (or live) cells represents the latest version of all common knowledge in CKB, and the set of history (or dead) cells represents all histories versions of common knowledge.
```
[the remark this answers is not in the document]
Comment (janx) [reply rfcs/0002-ckb/0002-ckb.md]: typo, fixed.

File: rfcs/0002-ckb/0002-ckb.md (hunk 3 above)
Comment (janx) [reply rfcs/0002-ckb/0002-ckb.md]: fixed.

Comment (janx) [reply rfcs/0002-ckb/0002-ckb.md]: updated.

**Pull request merged**